cranelift/frontend/src/frontend.rs

```diff
@@ -610,11 +630,186 @@ impl<'a> FunctionBuilder<'a> {
+        //    tightened up in the future, but we're starting with the easiest,
+        //    simplest thing. Besides, none of our mid-end optimization passes
+        //    have run at this point in time yet, so there probably isn't much,
+        //    if any, dead code.
+        for block in self.func_ctx.dfs.post_order_iter(&self.func) {
+            let mut option_inst = self.func.layout.last_inst(block);
+            while let Some(inst) = option_inst {
+                // (1) Remove values defined by this instruction from the `live`
+                // set.
+                for val in self.func.dfg.inst_results(inst) {
+                    live.remove(val);
+                }
+
+                // (2) If this instruction is a call, then we need to add a
+                // safepoint to record any values in `live`.
+                let opcode = self.func.dfg.insts[inst].opcode();
+                if opcode.is_call() && !opcode.is_return() {
+                    let mut live: SmallVec<[_; 4]> = live.iter().copied().collect();
+                    live.sort_by_key(|val| self.func.dfg.value_type(*val).bytes());
+                    for chunk in live.chunk_by(|a, b| {
+                        self.func.dfg.value_type(*a).bytes() == self.func.dfg.value_type(*b).bytes()
+                    }) {
+                        let byte_size = self.func.dfg.value_type(chunk[0]).bytes();
+                        debug_assert!(byte_size.is_power_of_two());
+                        let index = usize::try_from(byte_size.ilog2()).unwrap();
+                        max_vals_in_stack_map_by_size[index] =
+                            core::cmp::max(max_vals_in_stack_map_by_size[index], chunk.len());
+                    }
+
+                    let old_val = safepoints.insert(inst, live);
+                    debug_assert!(old_val.is_none());
+                }
+
+                // (3) Add all needs-stack-map values that are operands to this
+                // instruction to the live set.
+                for val in self.func.dfg.inst_values(inst) {
+                    if self.func_ctx.needs_stack_map.contains(val) {
+                        live.insert(val);
+                    }
+                }
+
+                option_inst = self.func.layout.prev_inst(inst);
+            }
+
+            // After we've processed this block's instructions, remove its
+            // parameters from the live set. This is part of step (1).
+            for val in self.func.dfg.block_params(block) {
+                live.remove(val);
+            }
+        }
+
+        // Create a stack slot for each size of needs-stack-map value. These
+        // slots are arrays capable of holding the maximum number of same-sized
+        // values that must appear in the same stack map at the same time.
+        //
+        // This is indexed by the log2 of the type size.
+        let mut stack_slots = [PackedOption::<ir::StackSlot>::default(); 5];
```

Comment:
Can we make the `5` a constant defined somewhere (and used in the initialization above too)?